File: slurm_term/sbatch_parser.py
"""Parse .sbatch files into Composer form state dicts.

Converts ``#SBATCH`` directives, ``module load`` lines, ``export`` lines,
and remaining shell commands into a dict compatible with
:meth:`ComposerTab.set_form_state`.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any


# Maps #SBATCH keys → form state keys
_DIRECTIVE_MAP: dict[str, str] = {
    "job-name": "name",
    "J":        "name",
    "partition": "partition",
    "p":        "partition",
    "time":     "time",
    "t":        "time",
    "nodes":    "nodes",
    "N":        "nodes",
    "ntasks-per-node": "ntasks",
    "cpus-per-task":   "cpus",
    "c":        "cpus",
    "mem":      "memory",
    "output":   "output",
    "o":        "output",
    "error":    "error",
    "e":        "error",
}

# These directives map to the GPU field but need value extraction
_GPU_DIRECTIVES = {"gres", "gpus", "gpus-per-node", "G"}

# Regex for #SBATCH lines: --key=value  or  --key value  or  -X value
_SBATCH_LONG_RE = re.compile(
    r"^\s*#SBATCH\s+--([a-zA-Z][a-zA-Z0-9_-]*)(?:=|\s+)(.+)?$"
)
_SBATCH_SHORT_RE = re.compile(
    r"^\s*#SBATCH\s+-([a-zA-Z])(?:\s+(.+))?$"
)

# Regex for module load lines
_MODULE_LOAD_RE = re.compile(r"^\s*module\s+load\s+(.+)$", re.IGNORECASE)

# Regex for export lines
_EXPORT_RE = re.compile(r"^\s*export\s+([A-Za-z_][A-Za-z0-9_]*=.+)$")


def parse_sbatch_file(path: str | Path) -> dict[str, Any]:
    """Parse an sbatch script file and return a Composer form state dict.

    Parameters
    ----------
    path : str or Path
        Absolute or relative path to the ``.sbatch`` file.

    Returns
    -------
    dict[str, Any]
        A dict with keys matching ``ComposerTab.set_form_state()`` format:
        ``mode``, ``name``, ``partition``, ``time``, ``nodes``, ``ntasks``,
        ``cpus``, ``memory``, ``gpus``, ``output``, ``error``, ``modules``,
        ``env``, ``init``.  An additional ``extra_directives`` key holds
        any unrecognised ``#SBATCH`` flags as a ``dict[str, str]``.

    Raises
    ------
    FileNotFoundError
        If *path* does not exist.
    ValueError
        If *path* is not a file or cannot be read.
    """
    p = Path(path)
    if not p.is_file():
        raise FileNotFoundError(f"File not found: {p}")

    try:
        text = p.read_text(encoding="utf-8", errors="replace")
    except OSError as e:
        raise ValueError(f"Cannot read file: {e}") from e
    return parse_sbatch_text(text)


def parse_sbatch_text(text: str) -> dict[str, Any]:
    """Parse raw sbatch script text into a Composer form state dict.

    This is the core parser; :func:`parse_sbatch_file` reads the file and
    delegates here.
    """
    state: dict[str, str] = {
        "mode": "sbatch",
        "name": "",
        "partition": "",
        "time": "",
        "nodes": "",
        "ntasks": "",
        "cpus": "",
        "memory": "",
        "gpus": "",
        "output": "",
        "error": "",
        "script": "",
    }
    extra_directives: dict[str, str] = {}
    modules: list[str] = []
    env_vars: list[str] = []
    init_cmds: list[str] = []

    # Track whether we've finished the directive block
    past_directives = False

    for line in text.splitlines():
        stripped = line.strip()

        # Skip shebang
        if stripped.startswith("#!"):
            continue

        # Skip pure comments (but not #SBATCH)
        if stripped.startswith("#") and not stripped.startswith("#SBATCH"):
            continue

        # Parse #SBATCH directives
        if stripped.startswith("#SBATCH"):
            m = _SBATCH_LONG_RE.match(stripped)
            if m:
                key, value = m.group(1), (m.group(2) or "").strip()
                _apply_directive(key, value, state, extra_directives)
                continue
            m = _SBATCH_SHORT_RE.match(stripped)
            if m:
                key, value = m.group(1), (m.group(2) or "").strip()
                _apply_directive(key, value, state, extra_directives)
                continue
            # Malformed #SBATCH line — skip
            continue

        # Empty lines before any commands
        if not stripped and not past_directives:
            continue

        past_directives = True

        # Skip empty lines in command section
        if not stripped:
            # Preserve blank lines in init commands for readability
            if init_cmds:
                init_cmds.append("")
            continue

        # Module loads
        m = _MODULE_LOAD_RE.match(stripped)
        if m:
            modules.append(m.group(1).strip())
            continue

        # Export lines
        m = _EXPORT_RE.match(stripped)
        if m:
            env_vars.append(m.group(1).strip())
            continue

        # Everything else is an init command
        init_cmds.append(line.rstrip())

    # Strip trailing blank lines from init_cmds
    while init_cmds and not init_cmds[-1].strip():
        init_cmds.pop()

    state["modules"] = "\n".join(modules)
    state["env"] = "\n".join(env_vars)
    state["init"] = "\n".join(init_cmds)

    result: dict[str, Any] = dict(state)
    if extra_directives:
        result["extra_directives"] = extra_directives
    return result


def _apply_directive(
    key: str,
    value: str,
    state: dict[str, str],
    extras: dict[str, str],
) -> None:
    """Apply a single #SBATCH directive to the state or extras dict."""
    # Check direct mapping first
    if key in _DIRECTIVE_MAP:
        form_key = _DIRECTIVE_MAP[key]
        state[form_key] = value
        return

    # GPU handling: extract count from gres=gpu:N or --gpus=N
    if key in _GPU_DIRECTIVES:
        gpu_val = value
        # Strip "gpu:" prefix if present (e.g. gres=gpu:a100:2 → a100:2)
        if gpu_val.lower().startswith("gpu:"):
            gpu_val = gpu_val[4:]
        state["gpus"] = gpu_val
        return

    # Unrecognised → store as extra directive
    extras[key] = value

File: src/sbatch_parser.rs
use std::collections::HashMap;

use regex::Regex;

/// Maps #SBATCH long keys to form state keys.
fn directive_map() -> HashMap<&'static str, &'static str> {
    let mut m = HashMap::new();
    m.insert("job-name", "name");
    m.insert("J", "name");
    m.insert("partition", "partition");
    m.insert("p", "partition");
    m.insert("time", "time");
    m.insert("t", "time");
    m.insert("nodes", "nodes");
    m.insert("N", "nodes");
    m.insert("ntasks-per-node", "ntasks");
    m.insert("cpus-per-task", "cpus");
    m.insert("c", "cpus");
    m.insert("mem", "memory");
    m.insert("output", "output");
    m.insert("o", "output");
    m.insert("error", "error");
    m.insert("e", "error");
    m
}

const GPU_DIRECTIVES: &[&str] = &["gres", "gpus", "gpus-per-node", "G"];

/// Parse raw sbatch script text into a form state dict.
pub fn parse_sbatch_text(text: &str) -> HashMap<String, String> {
    let dmap = directive_map();

    let long_re = Regex::new(r"^\s*#SBATCH\s+--([a-zA-Z][a-zA-Z0-9_-]*)(?:=|\s+)(.+)?$").unwrap();
    let short_re = Regex::new(r"^\s*#SBATCH\s+-([a-zA-Z])(?:\s+(.+))?$").unwrap();
    let module_re = Regex::new(r"(?i)^\s*module\s+load\s+(.+)$").unwrap();
    let export_re = Regex::new(r"^\s*export\s+([A-Za-z_][A-Za-z0-9_]*=.+)$").unwrap();

    let mut state: HashMap<String, String> = HashMap::new();
    state.insert("mode".into(), "sbatch".into());
    for key in &["name", "partition", "time", "nodes", "ntasks", "cpus",
                  "memory", "gpus", "output", "error", "script"] {
        state.insert((*key).into(), String::new());
    }

    let mut extra_directives: HashMap<String, String> = HashMap::new();
    let mut modules: Vec<String> = Vec::new();
    let mut env_vars: Vec<String> = Vec::new();
    let mut init_cmds: Vec<String> = Vec::new();
    let mut past_directives = false;

    for line in text.lines() {
        let stripped = line.trim();

        if stripped.starts_with("#!") {
            continue;
        }
        if stripped.starts_with('#') && !stripped.starts_with("#SBATCH") {
            continue;
        }

        if stripped.starts_with("#SBATCH") {
            if let Some(caps) = long_re.captures(stripped) {
                let key = caps.get(1).unwrap().as_str();
                let value = caps.get(2).map(|m| m.as_str().trim()).unwrap_or("");
                apply_directive(key, value, &dmap, &mut state, &mut extra_directives);
                continue;
            }
            if let Some(caps) = short_re.captures(stripped) {
                let key = caps.get(1).unwrap().as_str();
                let value = caps.get(2).map(|m| m.as_str().trim()).unwrap_or("");
                apply_directive(key, value, &dmap, &mut state, &mut extra_directives);
                continue;
            }
            continue;
        }

        if stripped.is_empty() && !past_directives {
            continue;
        }
        past_directives = true;

        if stripped.is_empty() {
            if !init_cmds.is_empty() {
                init_cmds.push(String::new());
            }
            continue;
        }

        if let Some(caps) = module_re.captures(stripped) {
            modules.push(caps.get(1).unwrap().as_str().trim().to_string());
            continue;
        }

        if let Some(caps) = export_re.captures(stripped) {
            env_vars.push(caps.get(1).unwrap().as_str().trim().to_string());
            continue;
        }

        init_cmds.push(line.trim_end().to_string());
    }

    // Strip trailing blanks from init_cmds
    while init_cmds.last().is_some_and(|l| l.trim().is_empty()) {
        init_cmds.pop();
    }

    state.insert("modules".into(), modules.join("\n"));
    state.insert("env".into(), env_vars.join("\n"));
    state.insert("init".into(), init_cmds.join("\n"));

    for (k, v) in &extra_directives {
        state.insert(format!("extra.{k}"), v.clone());
    }

    state
}

/// Parse a .sbatch file.
pub fn parse_sbatch_file(path: &str) -> Result<HashMap<String, String>, String> {
    let text = std::fs::read_to_string(path)
        .map_err(|e| format!("Cannot read file: {e}"))?;
    Ok(parse_sbatch_text(&text))
}

fn apply_directive(
    key: &str,
    value: &str,
    dmap: &HashMap<&str, &str>,
    state: &mut HashMap<String, String>,
    extras: &mut HashMap<String, String>,
) {
    if let Some(form_key) = dmap.get(key) {
        state.insert((*form_key).to_string(), value.to_string());
        return;
    }

    if GPU_DIRECTIVES.contains(&key) {
        let gpu_val = if let Some(rest) = value.strip_prefix("gpu:") {
            rest
        } else {
            value
        };
        state.insert("gpus".into(), gpu_val.to_string());
        return;
    }

    extras.insert(key.to_string(), value.to_string());
}
